Support for Redis Sentinel

XVincentX opened the pull request:
Connects #499 

## ioredis

### Mock library issues

- [x]  [The mock it's throwing exception when there's no data](https://github.com/stipsan/ioredis-mock/issues/328)
- [x] [The mock is missing implementation for `scan` command](https://github.com/stipsan/ioredis-mock/blob/master/compat.md), which we're using.
- [x] [The mock is missing support for reply transformers](https://github.com/stipsan/ioredis-mock#roadmap)

#### Solution

- Get rid of the `emulate` option and use Redis for everything, even for testing on CircleCI (we should do that anyway) https://github.com/ExpressGateway/express-gateway/pull/501/commits/4deacf977589ce74a95eba8e83813b0c18825eef

### Compatibility problems

- [x] All methods already return a Promise, there's no need for `promisify/Async` version https://github.com/ExpressGateway/express-gateway/pull/501/commits/5efe86733a1ce201ba8cec8a9785e1006f2cb0e6

- [x] `hgetall` reply must be modified to guarantee compatibility https://github.com/ExpressGateway/express-gateway/pull/501/commits/6c12cfdfc0324671a5f3752fedf71d67c0e7bfb0

- [x] `exec` function is using a Pipelining feature and has a different response type [fix](https://github.com/ExpressGateway/express-gateway/pull/501/commits/a58bca482f2edc8737b1d57c8d58f06e9b8f1eef#diff-1fe7eeca20bc9f4125db3c2909e5de55R186)

- [x] `bluebird` is overriding the default Promise implementation and this is preventing the `migrate` framework to work detect a Promise and react accordingly [issue](https://github.com/tj/node-migrate/issues/84) https://github.com/ExpressGateway/express-gateway/pull/501/commits/4ad3cfc4bdc23b68fca70fb45f3121b93249babf

XVincentX commented:
## Interesting followups:

1. Take advantage of [transparent key prefixing](https://github.com/luin/ioredis#transparent-key-prefixing) so we can remove some boilerplate code
2. Scan commands can return a stream — it is a great idea in order to avoid killing the server (both redis and the gateway)
3. [Cluster support](https://github.com/luin/ioredis#cluster). This is a different connection object with a different set of connection options as well.

DrMegavolt commented on test/db.test.js:
```diff
@@ -1,79 +0,0 @@
-const path = require('path');
```
this file was added to test ssl connection just recently why it is removed ?

XVincentX replied:
Good question.
I deleted it temporarily because I'll have to figure out how to test that part.

The problem here is — `ioredis` gets created using a constructor and not using a function such as `Redis.createClient()`

`sinon`, the stub library we're using, does not support spies on `constructors`.

I deleted this test just to verify it was working correctly. I'll now figure out another way to test this part.

DrMegavolt replied:
👍

DrMegavolt commented:
If the emulate option is working I'm thinking that we can land this massive PR, and break in memory a bit. so it will be another bug to fix it

XVincentX commented:
The emulate option is working correctly, yes.

I'm not really sure we should land this thing into master as it is.

I'm imagining somebody that wants to contribute or even simply trying the source code: they would clone the repository, install and run it: they would have a broke experience by default (as in-memory is not working and they would need a Redis instance), which is not very pleasant. I'd rather keep this branch here and resurrect it once `ioredis-mock` will be ready.

agonbina commented:
Just for the record, I am using(developing against) this branch to run EG and I haven't run into any issues when running with a single redis host or the sentinel mode.

DrMegavolt commented:
@agonbina thanks for active testing. the only reason we cannot merge it to master is some issues with InMemory mode. Which is implemented as mock of redis client. once fixed it will land into master

XVincentX commented:
I've updated the `ioredis-mock` package and rebased with latest master.

Two points still have to be solved — see [here](https://github.com/ExpressGateway/express-gateway/pull/501#issue-270940027)

XVincentX commented:
Rebase is going to be fun.

DrMegavolt commented:
squash manually, it will be way easier.

DrMegavolt commented:
LGTM

agonbina commented:
Nice to see this was merged! I will switch to using the master branch now 👍 I'm curious as to when is the next release planned?

XVincentX commented:
Hard work, but we made it!

These two images will summarize last couple of hours here:

![image](https://user-images.githubusercontent.com/1416224/32864219-8a35b32a-ca5e-11e7-8592-6de2961f2da3.png)
![firecomputer](https://user-images.githubusercontent.com/1416224/32864230-96c23924-ca5e-11e7-9e59-4ab1359a36ca.png)

DrMegavolt commented:
@agonbina waiting for multi-auth merge. 
Expected Thursday Nov 16. If something goes wrong then next week
